### PR TITLE
Add code activities to the ch5 (behat-hooks-background)

### PR DESCRIPTION
**Pull request by bocharsky-bw**
Resolve `05-behat-hooks-background` in the #12

### CODE REVIEW

File: knpu/Challenges/BehatHooksBackground/CheckStepTestResultCoding.php [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+EOF
+            );
+        }
+    }
+
+    public function grade(CodingExecutionResult $result)
+    {
+        $featureContextClass = new \ReflectionClass('FeatureContext');
+        if (!$featureContextClass->hasMethod('afterStepHook')) {
+            throw new GradingException('Method `afterStepHook()` doesn\'t found in the `FeatureContext` class. Did you create it?');
```
Comment (weaverryan) [inline knpu/Challenges/BehatHooksBackground/CheckStepTestResultCoding.php]: How about:

The `afterStepHook()` method wasn't found in the `FeatureContext` class.

File: knpu/Challenges/BehatHooksBackground/afterStepHookCoding.php [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+        if (!$featureContextClass->hasMethod('afterStepHook')) {
+            throw new GradingException('Method `afterStepHook()` doesn\'t found in the `FeatureContext` class. Did you create it?');
+        }
+        if (!$featureContextClass->getMethod('afterStepHook')->isPublic()) {
+            throw new GradingException('Method `afterStepHook()` should be public.');
+        }
+        $docComment = $featureContextClass->getMethod('afterStepHook')->getDocComment();
+        if (false === strpos($docComment, '@AfterStep')) {
+            throw new GradingException('You should to use `@AfterStep` annotation for `afterStepHook()` method.');
+        }
```
Comment (weaverryan) [inline knpu/Challenges/BehatHooksBackground/afterStepHookCoding.php:134]: nice grading :+1:

File: knpu/Challenges/BehatHooksBackground/CheckStepTestResultCoding.php [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+    /**
+     * @return string
+     */
+    public function getQuestion()
+    {
+        return <<<EOF
+Whenever you have a hook function, you're actually passed an `\$event`
+argument object that contains information about what's happening inside Behat
+at this moment. The exact object depends on which hook you're using
+(see [Behat Hooks](http://docs.behat.org/en/v3.0/guides/3.hooks.html#hooks)].
```
Comment (weaverryan) [inline knpu/Challenges/BehatHooksBackground/CheckStepTestResultCoding.php]: one extra `]` at the end

File: knpu/Challenges/BehatHooksBackground/CheckStepTestResultCoding.php [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+            throw new GradingException('Method `afterStepHook()` should be public.');
+        }
+        $afterStepHookMethod = $featureContextClass->getMethod('afterStepHook');
+        $docComment = $afterStepHookMethod->getDocComment();
+        if (false === strpos($docComment, '@AfterStep')) {
+            throw new GradingException('You should to use `@AfterStep` annotation for `afterStepHook()` method.');
+        }
+        $afterStepHookParameters = $afterStepHookMethod->getParameters();
+        if (1 !== $afterStepHookMethod->getNumberOfRequiredParameters()) {
+            throw new GradingException('Method `afterStepHook()` should get one required parameter.');
```
Comment (weaverryan) [inline knpu/Challenges/BehatHooksBackground/CheckStepTestResultCoding.php]: how about:

> Make sure you give the `afterStepHook()` method exactly one argument.

File: knpu/Challenges/BehatHooksBackground/CheckStepTestResultCoding.php [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+        }
+        if (0 !== strcmp('event', $afterStepHookParameters[0]->getName())) {
+            throw new GradingException('First argument in the `afterStepHook()` method should be named `$event`.');
```
Comment (weaverryan) [inline knpu/Challenges/BehatHooksBackground/CheckStepTestResultCoding.php]: I want to be clear that _we_ want it to be called `$event`, but it doesn't need to be for functionality purposes:

``` rst
Though you can really call it anything, let's call the argument to `afterStepHook()`
`$event` for clarity.
```

File: knpu/Challenges/BehatHooksBackground/CheckStepTestResultCoding.php (hunk 3 above)
Comment (bocharsky-bw) [inline knpu/Challenges/BehatHooksBackground/CheckStepTestResultCoding.php]: Thanks! Didn't see it in activity runner

Comment (weaverryan) [inline knpu/Challenges/BehatHooksBackground/CheckStepTestResultCoding.php]: makes sense - runner doesn't render markdown :)

File: knpu/Challenges/BehatHooksBackground/CheckStepTestResultCoding.php (hunk 1 above)
Comment (bocharsky-bw) [inline knpu/Challenges/BehatHooksBackground/CheckStepTestResultCoding.php]: I think it's better :+1:  I just keep it consistent with  `oo` tut. We start message with `Method methodName() ...` there.

File: knpu/Challenges/BehatHooksBackground/CheckStepTestResultCoding.php (hunk 4 above)
Comment (bocharsky-bw) [inline knpu/Challenges/BehatHooksBackground/CheckStepTestResultCoding.php]: :+1:

File: knpu/Challenges/BehatHooksBackground/CheckStepTestResultCoding.php (hunk 5 above)
Comment (bocharsky-bw) [inline knpu/Challenges/BehatHooksBackground/CheckStepTestResultCoding.php]: :+1: I also wondering should we check somehow that student call `$event->getTestResult()->isPassed()` and dump result in `afterStepHook()` method?

**Pull request merged**